### ISSUE TITLE
Adding Q-DEIM

### ISSUE THREAD

**Issue by pmli** [pymor/pymor#730]
It would be nice for pyMOR to have Q-DEIM (see section 2 and Matlab code at the end of page A639 in [1]). Unfortunately, as far as I can tell, it is not possible to implement it using pyMOR's interfaces.

If `U` is a `VectorArray`, thought of as a matrix of column vectors, representing the result of POD applied to the snapshots of the nonlinear operator, the Q-DEIM algorithm consists mainly of a QR decomposition with column pivoting applied to `U^H` (or LQ decomposition with row pivoting to `U`). The first step is finding a row in `U` with the largest Euclidian norm, which I don't see how to do using pyMOR's interfaces.

Another option might be implementing randomized Q-DEIM (Q-DEIMr, see section 3 in [1]), which only works with a (small) subset of the rows of `U`. This should then allow using `dofs` method.

[1] Z. Drmač and S. Gugercin, [A New Selection Operator for the Discrete Empirical Interpolation Method---Improved A Priori Error Bound and Extensions](https://doi.org/10.1137/15M1019271)

**Comment by sdrave** [comment]
I had a glance at [1]. Do I understand correctly that the only thing preventing us from implementing Q-DEIM is that we cannot do the mentioned Businger-Golub pivoting through the `VectorArrayInterface`? Everything else should work since we are computing the pivoted QR decomposition of a matrix given by a `VectorArray` `U` of *row* vectors. Left-multiplying `U` by Householder matrices should doable via `lincomb` and keeping track of the column permutation should be possible as well. - If that's all true, maybe there is another pivoting strategy one could use?

Apart from that, if I understood the paper correctly, only computing `S` is computationally feasible, whereas `S_*`, which has a much better a priori estimate, is not computable? Is there work to approximately compute `S_*`?

**Comment by pmli** [comment]
From what I understand:
- finding `S_*` is a combinatorial problem (might be NP hard)
- computing `S` by Businger-Golub QR with column pivoting (QRP) can be thought of as a greedy approach for approximating `S_*` (remark 2.3 in [1])
- the exponential term in QRP's error bound cannot be removed, as shown by Kahan matrices example
- in practice, QRP behaves much better than what the error bound would suggest (remark 2.2 [1])
- using strong rank-revealing QR (SRRQR) gives an error bound similar to the one for `S_*` (see [2] or [3])
- SRRQR is more expensive than QRP (and also uses Businger-Golub pivoting)

I'm not aware of an approach not using Businger-Golub pivoting.

[2] Z. Drmač, A. K. Saibaba, [The Discrete Empirical Interpolation Method: Canonical Structure and Formulation in Weighted Inner Product Spaces](https://doi.org/10.1137/17M1129635)
[3] Z. Drmač, [DEIM slides](http://www.math.rug.nl/gcsc/files/drmac_slides_part2.pdf), from [Groningen MOR School](http://www.math.rug.nl/gcsc/morschool.html)

**Comment by sdrave** [comment]
@pymor/pymor-devs, anyone interested in working on this for the next realease. I think it would be really nice to have it in pyMOR ..

**Comment by pmli** [comment]
Did we decide how to deal with computing l2 norms of all the dofs? Or instead to go for implementing randomized Q-DEIM?

**Comment by sdrave** [comment]
I think we agreed to implement #765 and #764.

**Comment by pmli** [comment]
Ok, I thought we were looking for more use cases. Anyway, at least #765 needs to be implemented, maybe also #764 to force exact zeros in Householder reflections.

**Comment by sdrave** [comment]
> Ok, I thought we were looking for more use cases.

I remember that to be a pre-dev meeting state, but I could be wrong ..

> Anyway, at least #765 needs to be implemented, 

Yes. Unfortunately, this will get a bit complicated, as we need to decide to to actually specify the function that is to be applied.

> maybe also #764 to force exact zeros in Householder reflections.

That, on the other hand, should be fairly easy to realize.

**Comment by pmli** [comment]
> Yes. Unfortunately, this will get a bit complicated, as we need to decide to to actually specify the function that is to be applied.

Right, I guess having a `VectorArray.apply` method would be a bit confusing.

How about we continue the discussions in #765 and #764? Then we could return to the point of implementing Q-DEIM?

**Comment by sdrave** [comment]
Done in #2493.
